marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java

```diff
@@ -1,477 +1,184 @@
-        this.flowItemFailureListeners.add(listener);
-        return this;
-    }
+        //Put response, steps and flow in maps with jobId as key
+        flowResp.put(jobId, response);
+        stepsMap.put(jobId, stepsQueue);
+        flowMap.put(jobId, flow);
 
-    @Override
-    public FlowRunner onStatusChanged(FlowStatusListener listener) {
-        this.flowStatusListeners.add(listener);
-        return this;
-    }
+        //add jobId to a queue
+        jobQueue.add(jobId);
 
-    @Override
-    public FlowRunner onFinished(FlowFinishedListener listener) {
-        this.flowFinishedListeners.add(listener);
-        return this;
-    }
-
-    @Override
-    public void awaitCompletion() {
-        try {
-            awaitCompletion(Long.MAX_VALUE, TimeUnit.DAYS);
-        } catch (InterruptedException| TimeoutException e) {
+        if(! isRunning.get()){
+            initializeFlow(jobId);
         }
+        return response;
     }
 
-    @Override
-    public void awaitCompletion(long timeout, TimeUnit unit) throws InterruptedException,TimeoutException {
-        if (runningThread != null) {
-            runningThread.join(unit.convert(timeout, unit));
-            if (runningThread.getState() != Thread.State.TERMINATED) {
-                if ( dataMovementManager != null && queryBatcher != null ) {
-                    dataMovementManager.stopJob(queryBatcher);
-                }
-                runningThread.interrupt();
-                throw new TimeoutException("Timeout occurred after "+timeout+" "+unit.toString());
-            }
+    private void initializeFlow(String jobId) {
+        isRunning.set(true);
+        runningJobId = jobId;
+        runningFlow = flowMap.get(runningJobId);
+        if(threadPool == null || threadPool.isTerminated()) {
+            threadPool = Executors.newFixedThreadPool(1);
         }
+        threadPool.submit(new FlowRunnerTask(runningFlow, runningJobId));
     }
 
-    @Override
-    public Job run() {
-        runningThread = null;
-        Job job = createJob();
-
-        if (options == null) {
-            options = new HashMap<>();
-        } else {
-            if (options.get("fullOutput") != null) {
-                isFullOutput = Boolean.parseBoolean(options.get("fullOutput").toString());
-            }
-        }
-        options.put("flow", this.flow.getName());
-        Collection<String> uris = null;
-        try {
-            uris = runCollector();
-        } catch (Exception e) {
-            job.setCounts(0,0, 0, 0, 0)
-                .withStatus(JobStatus.FAILED.toString());
-            StringWriter errors = new StringWriter();
-            e.printStackTrace(new PrintWriter(errors));
-            job.withJobOutput(errors.toString());
-            jobUpdate.postJobs(jobId, JobStatus.FAILED.toString(), step);
-            return job;
+    public void stopJob(String jobId) {
+        if (jobId.equals(runningJobId)) {
+            stepsMap.get(jobId).clear();
```

Comment:
clear() is only available on the map
```suggestion
            stepsMap.get(jobId).clear();
            stepsMap.remove(jobId);
```